src/CHIPDriver.hh

```diff
@@ -118,4 +120,171 @@ extern std::atomic_ulong CHIPNumRegisteredFatBinaries;
  */
 extern thread_local hipError_t CHIPTlsLastError;
 
+class DeviceType {
+public:
+  enum Type { GPU, CPU, ACCEL, FPGA, DEFAULT };
+
+private:
+  Type Type_;
+
+public:
+  DeviceType(Type TypeIn) : Type_(TypeIn) {}
+
+  std::string str() const {
+    switch (Type_) {
+    case GPU:
+      return "gpu";
+    case CPU:
+      return "cpu";
+    case ACCEL:
+      return "accel";
+    case FPGA:
+      return "fpga";
+    case DEFAULT:
+      return "default";
+    default:
+      return "unknown";
+    }
+  }
+
+  Type getType() const { return Type_; }
+};
+
+class BackendType {
+public:
+  enum Type { OPENCL, LEVEL0, DEFAULT };
+
+private:
+  Type Type_;
+
+public:
+  BackendType(Type TypeIn) : Type_(TypeIn) {}
+
+  const char *str() const {
+    switch (Type_) {
+    case OPENCL:
+      return "opencl";
+    case LEVEL0:
+      return "level0";
+    case DEFAULT:
+      return "default";
+    default:
+      return "unknown";
+    }
+  }
+
+  Type getType() const { return Type_; }
+};
+
+class EnvVars {
+public:
+  int PlatformIdx = 0;
+  DeviceType Device = DeviceType::GPU;
+  int DeviceIdx = 0;
+  BackendType Backend = BackendType::DEFAULT;
+  bool DumpSpirv = false;
+  std::string JitFlags = CHIP_DEFAULT_JIT_FLAGS;
+  bool L0ImmCmdLists = true;
+  int L0CollectEventsTimeout = 0;
```

Comment:
Should these be immutable after environment variables are parsed?

Reply:
Making them const would complicate constructor but I'll make them private.

Reply:
Sorry, I meant immutability respect to class users. Making them private also makes them immutable for the users.

Reply:
So does this current solution work for you then?